lib/remap-functions/src/parse_timestamp.rs

```diff
@@ -21,106 +21,56 @@ impl Function for ParseTimestamp {
 }
 
 impl ParseTimestampFn {
     #[cfg(test)]
-    fn new(format: &str, value: Box<dyn Expression>, default: Option<Value>) -> Self {
+    fn new(format: &str, value: Box<dyn Expression>) -> Self {
         let format = Box::new(Literal::from(format));
-        let default = default.map(|v| Box::new(Literal::from(v)) as _);
 
-        Self {
-            value,
-            format,
-            default,
-        }
+        Self { value, format }
     }
 }
 
 impl Expression for ParseTimestampFn {
     fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let value = self.value.execute(state, object)?;
         let format = self.format.execute(state, object);
 
-        let to_timestamp = |value| match value {
+        match value {
             Value::Bytes(v) => format
-                .clone()
                 .map(|v| format!("timestamp|{}", String::from_utf8_lossy(&v.unwrap_bytes())))?
                 .parse::<Conversion>()
                 .map_err(|e| format!("{}", e))?
                 .convert(v)
                 .map_err(|e| e.to_string().into()),
             Value::Timestamp(_) => Ok(value),
             _ => Err("unable to convert value to integer".into()),
-        };
-
-        crate::util::convert_value_or_default(
-            self.value.execute(state, object),
-            self.default.as_ref().map(|v| v.execute(state, object)),
-            to_timestamp,
-        )
+        }
     }
 
     fn type_def(&self, state: &state::Compiler) -> TypeDef {
-        let value_def = self
-            .value
-            .type_def(state)
-            .fallible_unless(value::Kind::Timestamp);
-
-        let default_def = self
-            .default
-            .as_ref()
-            .map(|v| v.type_def(state).fallible_unless(value::Kind::Timestamp));
-
-        // The `format` type definition is only relevant if:
-        //
-        // 1. `value` can resolve to a string, AND:
-        //   1. `default` is not defined, OR
-        //   2. `default` can also resolve to a string.
-        //
-        // The `format` type is _always_ fallible, because its string has to be
-        // parsed into a valid timestamp format.
-        let format_def = if value_def.kind.contains(value::Kind::Bytes) {
-            match &default_def {
-                Some(def) if def.kind.contains(value::Kind::Bytes) => {
-                    Some(self.format.type_def(state).into_fallible(true))
-                }
-                Some(_) => None,
-                None => Some(self.format.type_def(state).into_fallible(true)),
-            }
-        } else {
-            None
-        };
+        use value::Kind;
 
-        value_def
-            .merge_with_default_optional(default_def)
-            .merge_optional(format_def)
-            .with_constraint(value::Kind::Timestamp)
+        self.value
+            .type_def(state)
+            .fallible_unless(Kind::Bytes | Kind::Timestamp)
```

Comment:
I believe this function is always fallible, since we have to parse the `format` string at runtime.